{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Initialization"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "R version 4.4.1 (2024-06-14 ucrt)\n",
       "Platform: x86_64-w64-mingw32/x64\n",
       "Running under: Windows 11 x64 (build 22631)\n",
       "\n",
       "Matrix products: default\n",
       "\n",
       "\n",
       "locale:\n",
       "[1] LC_COLLATE=English_United States.utf8 \n",
       "[2] LC_CTYPE=English_United States.utf8   \n",
       "[3] LC_MONETARY=English_United States.utf8\n",
       "[4] LC_NUMERIC=C                          \n",
       "[5] LC_TIME=English_United States.utf8    \n",
       "\n",
       "time zone: America/New_York\n",
       "tzcode source: internal\n",
       "\n",
       "attached base packages:\n",
       "[1] grid      stats4    stats     graphics  grDevices utils     datasets \n",
       "[8] methods   base     \n",
       "\n",
       "other attached packages:\n",
       " [1] gprofiler2_0.2.3            VennDiagram_1.7.3          \n",
       " [3] futile.logger_1.4.3         UCell_2.8.0                \n",
       " [5] pheatmap_1.0.12             msigdbr_7.5.1              \n",
       " [7] enrichplot_1.24.4           org.Mm.eg.db_3.19.1        \n",
       " [9] AnnotationDbi_1.66.0        clusterProfiler_4.13.0     \n",
       "[11] pathview_1.44.0             DOSE_3.30.5                \n",
       "[13] gridExtra_2.3               SingleR_2.6.0              \n",
       "[15] celldex_1.14.0              DESeq2_1.44.0              \n",
       "[17] SummarizedExperiment_1.34.0 Biobase_2.64.0             \n",
       "[19] MatrixGenerics_1.16.0       matrixStats_1.4.1          \n",
       "[21] GenomicRanges_1.56.1        GenomeInfoDb_1.40.1        \n",
       "[23] IRanges_2.38.1              S4Vectors_0.42.1           \n",
       "[25] BiocGenerics_0.50.0         scCustomize_2.1.2          \n",
       "[27] scales_1.3.0                circlize_0.4.16            \n",
       "[29] scRepertoire_2.0.5          ggrepel_0.9.6              \n",
       "[31] viridis_0.6.5               viridisLite_0.4.2          \n",
       "[33] patchwork_1.3.0             lubridate_1.9.3            \n",
       "[35] forcats_1.0.0               purrr_1.0.2                \n",
       "[37] readr_2.1.5                 tidyr_1.3.1                \n",
       "[39] tibble_3.2.1                tidyverse_2.0.0            \n",
       "[41] Stat2Data_2.0.0             stringr_1.5.1              \n",
       "[43] ggplot2_3.5.1               cowplot_1.1.3              \n",
       "[45] dplyr_1.1.4                 Seurat_5.1.0               \n",
       "[47] SeuratObject_5.0.2          sp_2.1-4                   \n",
       "\n",
       "loaded via a namespace (and not attached):\n",
       "  [1] R.methodsS3_1.8.2           goftest_1.2-3              \n",
       "  [3] Biostrings_2.72.1           HDF5Array_1.32.1           \n",
       "  [5] vctrs_0.6.5                 spatstat.random_3.3-2      \n",
       "  [7] digest_0.6.37               png_0.1-8                  \n",
       "  [9] shape_1.4.6.1               gypsum_1.0.1               \n",
       " [11] IRdisplay_1.1               deldir_2.0-4               \n",
       " [13] parallelly_1.38.0           MASS_7.3-61                \n",
       " [15] reshape2_1.4.4              httpuv_1.6.15              \n",
       " [17] qvalue_2.36.0               withr_3.0.1                \n",
       " [19] ggrastr_1.0.2               ggfun_0.1.6                \n",
       " [21] survival_3.7-0              memoise_2.0.1              \n",
       " [23] ggbeeswarm_0.7.2            gson_0.1.0                 \n",
       " [25] janitor_2.2.0               MatrixModels_0.5-3         \n",
       " [27] tidytree_0.4.6              zoo_1.8-12                 \n",
       " [29] GlobalOptions_0.1.2         KEGGgraph_1.64.0           \n",
       " [31] pbapply_1.7-2               IRkernel_1.3.2             \n",
       " [33] R.oo_1.26.0                 rematch2_2.1.2             \n",
       " [35] KEGGREST_1.44.1             promises_1.3.0             \n",
       " [37] evmix_2.12                  httr_1.4.7                 \n",
       " [39] globals_0.16.3              fitdistrplus_1.2-1         \n",
       " [41] rhdf5filters_1.16.0         rhdf5_2.48.0               \n",
       " [43] UCSC.utils_1.0.0            miniUI_0.1.1.1             \n",
       " [45] generics_0.1.3              ggalluvial_0.12.5          \n",
       " [47] base64enc_0.1-3             babelgene_22.9             \n",
       " [49] curl_5.2.3                  repr_1.1.7                 \n",
       " [51] zlibbioc_1.50.0             ScaledMatrix_1.12.0        \n",
       " [53] ggraph_2.2.1                polyclip_1.10-7            \n",
       " [55] GenomeInfoDbData_1.2.12     ExperimentHub_2.12.0       \n",
       " [57] SparseArray_1.4.8           xtable_1.8-4               \n",
       " [59] evaluate_1.0.0              S4Arrays_1.4.1             \n",
       " [61] BiocFileCache_2.12.0        hms_1.1.3                  \n",
       " [63] irlba_2.3.5.1               colorspace_2.1-1           \n",
       " [65] filelock_1.0.3              ROCR_1.0-11                \n",
       " [67] reticulate_1.39.0           spatstat.data_3.1-2        \n",
       " [69] Rgraphviz_2.48.0            magrittr_2.0.3             \n",
       " [71] lmtest_0.9-40               snakecase_0.11.1           \n",
       " [73] ggtree_3.12.0               later_1.3.2                \n",
       " [75] lattice_0.22-6              spatstat.geom_3.3-3        \n",
       " [77] future.apply_1.11.2         SparseM_1.84-2             \n",
       " [79] shadowtext_0.1.4            scattermore_1.2            \n",
       " [81] XML_3.99-0.17               RcppAnnoy_0.0.22           \n",
       " [83] pillar_1.9.0                nlme_3.1-166               \n",
       " [85] compiler_4.4.1              beachmat_2.20.0            \n",
       " [87] RSpectra_0.16-2             stringi_1.8.4              \n",
       " [89] tensor_1.5                  plyr_1.8.9                 \n",
       " [91] crayon_1.5.3                abind_1.4-8                \n",
       " [93] ggdendro_0.2.0              gridGraphics_0.5-1         \n",
       " [95] locfit_1.5-9.10             org.Hs.eg.db_3.19.1        \n",
       " [97] graphlayouts_1.2.0          bit_4.5.0                  \n",
       " [99] fastmatch_1.1-4             codetools_0.2-20           \n",
       "[101] BiocSingular_1.20.0         alabaster.ranges_1.4.2     \n",
       "[103] paletteer_1.6.0             plotly_4.10.4              \n",
       "[105] mime_0.12                   splines_4.4.1              \n",
       "[107] Rcpp_1.0.12                 fastDummies_1.7.4          \n",
       "[109] quantreg_5.98               dbplyr_2.5.0               \n",
       "[111] sparseMatrixStats_1.16.0    blob_1.2.4                 \n",
       "[113] utf8_1.2.4                  BiocVersion_3.19.1         \n",
       "[115] pbdZMQ_0.3-13               fs_1.6.4                   \n",
       "[117] listenv_0.9.1               evd_2.3-7.1                \n",
       "[119] DelayedMatrixStats_1.26.0   ggplotify_0.1.2            \n",
       "[121] gsl_2.1-8                   Matrix_1.7-0               \n",
       "[123] tzdb_0.4.0                  tweenr_2.0.3               \n",
       "[125] pkgconfig_2.0.3             tools_4.4.1                \n",
       "[127] cachem_1.1.0                RSQLite_2.3.7              \n",
       "[129] DBI_1.2.3                   fastmap_1.2.0              \n",
       "[131] ica_1.0-3                   AnnotationHub_3.12.0       \n",
       "[133] ggprism_1.0.5               BiocManager_1.30.25        \n",
       "[135] dotCall64_1.1-1             graph_1.82.0               \n",
       "[137] RANN_2.6.2                  alabaster.schemas_1.4.0    \n",
       "[139] farver_2.1.2                scatterpie_0.2.4           \n",
       "[141] tidygraph_1.3.1             yaml_2.3.10                \n",
       "[143] VGAM_1.1-12                 cli_3.6.3                  \n",
       "[145] leiden_0.4.3.1              lifecycle_1.0.4            \n",
       "[147] uwot_0.2.2                  lambda.r_1.2.4             \n",
       "[149] BiocParallel_1.38.0         timechange_0.3.0           \n",
       "[151] gtable_0.3.5                rjson_0.2.23               \n",
       "[153] ggridges_0.5.6              progressr_0.14.0           \n",
       "[155] ape_5.8                     cubature_2.1.1             \n",
       "[157] parallel_4.4.1              jsonlite_1.8.8             \n",
       "[159] bitops_1.0-8                RcppHNSW_0.6.0             \n",
       "[161] bit64_4.5.2                 Rtsne_0.17                 \n",
       "[163] yulab.utils_0.1.7           BiocNeighbors_1.22.0       \n",
       "[165] alabaster.matrix_1.4.2      spatstat.utils_3.1-0       \n",
       "[167] futile.options_1.0.1        alabaster.se_1.4.1         \n",
       "[169] GOSemSim_2.30.2             spatstat.univar_3.0-1      \n",
       "[171] R.utils_2.12.3              truncdist_1.0-2            \n",
       "[173] lazyeval_0.2.2              alabaster.base_1.4.2       \n",
       "[175] shiny_1.9.1                 htmltools_0.5.8.1          \n",
       "[177] GO.db_3.19.1                iNEXT_3.0.1                \n",
       "[179] sctransform_0.4.1           formatR_1.14               \n",
       "[181] rappdirs_0.3.3              glue_1.7.0                 \n",
       "[183] spam_2.10-0                 httr2_1.0.5                \n",
       "[185] XVector_0.44.0              RCurl_1.98-1.16            \n",
       "[187] treeio_1.28.0               igraph_2.0.3               \n",
       "[189] R6_2.5.1                    SingleCellExperiment_1.26.0\n",
       "[191] cluster_2.1.6               Rhdf5lib_1.26.0            \n",
       "[193] stringdist_0.9.12           aplot_0.2.3                \n",
       "[195] DelayedArray_0.30.1         tidyselect_1.2.1           \n",
       "[197] vipor_0.4.7                 ggforce_0.4.2              \n",
       "[199] future_1.34.0               rsvd_1.0.5                 \n",
       "[201] munsell_0.5.1               KernSmooth_2.23-24         \n",
       "[203] data.table_1.15.4           htmlwidgets_1.6.4          \n",
       "[205] fgsea_1.30.0                RColorBrewer_1.1-3         \n",
       "[207] rlang_1.1.4                 spatstat.sparse_3.1-0      \n",
       "[209] spatstat.explore_3.3-2      uuid_1.2-1                 \n",
       "[211] fansi_1.0.6                 beeswarm_0.4.0             "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# rm(list = ls())\n",
    "# #Library the required\n",
    "\n",
    "# suppressMessages(library(Seurat))\n",
    "# suppressMessages(library(dplyr))\n",
    "# suppressMessages(library(cowplot))\n",
    "# suppressMessages(library(ggplot2))\n",
    "\n",
    "# suppressMessages(library(stringr))\n",
    "# suppressMessages(library(Stat2Data))\n",
    "# suppressMessages(library(tidyverse))\n",
    "# suppressMessages(library(patchwork))\n",
    "# # library(pheatmap))\n",
    "# suppressMessages(library(viridis)) \n",
    "# suppressMessages(library(ggplot2))\n",
    "# suppressMessages(library(ggrepel))\n",
    "# suppressMessages(library(scRepertoire))\n",
    "# suppressMessages(library(circlize))\n",
    "# suppressMessages(library(scales))\n",
    "# suppressMessages(library(scCustomize))\n",
    "# suppressMessages(library(DESeq2)) \n",
    "# suppressMessages(library(celldex))\n",
    "# suppressMessages(library(SingleR))\n",
    "# suppressMessages(library(gridExtra))\n",
    "\n",
    "# suppressMessages(library(DOSE))\n",
    "# suppressMessages(library(pathview))\n",
    "# suppressMessages(library(clusterProfiler))\n",
    "# suppressMessages(library(org.Mm.eg.db))\n",
    "# suppressMessages(library(enrichplot))\n",
    "# suppressMessages(library(msigdbr))\n",
    "# suppressMessages(library(pheatmap))\n",
    "# suppressMessages(library(UCell))\n",
    "# suppressMessages(library(VennDiagram))\n",
    "# suppressMessages(library(gprofiler2))\n",
    "\n",
    "# sessionInfo()\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Overrepresentation analysis function GO and MSigDB -- ClusterProfiler"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GO_overrepresentation_analysis <- function (significant_genes, all_genes, local_path, ontology = 'ALL', minGSSize = 5, maxGSSize = 500, filename = '') {\n",
    "\n",
    "     color_scale <- viridis(n = 4, direction = -1)\n",
    "     options(enrichplot.colours = color_scale)\n",
    "     \n",
    "     enrichment_results <- enrichGO(gene = significant_genes, \n",
    "                    universe = all_genes,\n",
    "                    keyType = \"SYMBOL\",\n",
    "                    OrgDb = org.Mm.eg.db, \n",
    "                    ont = ontology, \n",
    "                    pAdjustMethod = \"BH\", \n",
    "                    minGSSize    = minGSSize,\n",
    "                    maxGSSize    = maxGSSize,\n",
    "                    qvalueCutoff = 0.25)\n",
    "\n",
    "     enrichment_results_table <- as_tibble(enrichment_results)\n",
    "     write.csv(enrichment_results_table, paste0(local_path, filename,'GO_OverRepresentation_analysis_results_', ontology, '.csv'))\n",
    "\n",
    "     if (nrow(enrichment_results_table) > 1) {\n",
    "          ## Add similarity matrix to the termsim slot of enrichment result\n",
    "          enrichment_results <- enrichplot::pairwise_termsim(enrichment_results)\n",
    "          enrichment_results <- simplify(enrichment_results, cutoff=0.7, by=\"p.adjust\", select_fun=min)\n",
    "\n",
    "          dotplot(enrichment_results,\n",
    "               showCategory=50,\n",
    "               title = paste0(filename,'GO Overrepresentation analysis ', ontology),\n",
    "               label_format = 60)\n",
    "          ggsave(paste0(filename, 'GO overrepresentation_analysis_dotplot_', ontology,'.pdf'), width = 10, height = 18, path = local_path)\n",
    "\n",
    "          ## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms\n",
    "          try(emapplot(enrichment_results, showCategory = (nrow(enrichment_results_table-1)),\n",
    "               cex.params = list(category_node = 0.8, line = 0.3, category_label = 1)) + ggtitle(paste0(filename, 'Overrepresentation analysis ', ontology)))\n",
    "          ggsave(paste0(filename, 'GO_overrepresentation_analysis_network_', ontology,'.pdf'), width = 14, height = 18, path = local_path)\n",
    "     }\n",
    "}\n",
    "\n",
    "GO_GSEA_analysis <- function (results, local_path, ontology = 'ALL') {\n",
    "     \n",
    "     color_scale <- viridis(n = 4, direction = -1)\n",
    "     options(enrichplot.colours = color_scale)\n",
    "\n",
    "          #### GSEA ####\n",
    "\n",
    "     fold_changes <- results |> arrange(desc(log2FoldChange)) |> pull(log2FoldChange)\n",
    "     names(fold_changes) <- results |> arrange(desc(log2FoldChange)) |> pull(genes)\n",
    "\n",
    "     gsea_results <- gseGO(geneList     = fold_changes,\n",
    "               OrgDb        = org.Mm.eg.db,\n",
    "               ont          = ontology,\n",
    "               keyType = \"SYMBOL\",\n",
    "               minGSSize    = 5,\n",
    "               maxGSSize    = 500,\n",
    "               pvalueCutoff = 0.05,\n",
    "               verbose      = FALSE)               \n",
    "\n",
    "     gsea_results_table <- as_tibble(gsea_results) \n",
    "     write.csv(gsea_results_table, paste0(local_path, 'Gene_Set_Enrichment_Analysis_results_', ontology,'.csv'))\n",
    "\n",
    "     if (nrow(gsea_results_table) > 1) {\n",
    "          dotplot(gsea_results,\n",
    "               showCategory=50,\n",
    "               title = paste0('GSEA analysis ', ontology),\n",
    "               label_format = 60)\n",
    "          ggsave(paste0('GSEA_dotplot_', ontology,'.pdf'), width = 10, height = 18, path = local_path)\n",
    "\n",
    "          ## Add similarity matrix to thenes,  termsim slot of enrichment result\n",
    "          gsea_results <- enrichplot::pairwise_termsim(gsea_results)\n",
    "\n",
    "          ## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms\n",
    "          try(emapplot(gsea_results, showCategory = 60,\n",
    "               cex.params = list(category_node = 0.8, line = 0.3, category_label = 1)) + ggtitle(paste0('GSEA analysis ', ontology)))\n",
    "          ggsave(paste0('GSEA_network_', ontology,'.pdf'), width = 14, height = 18, path = local_path)\n",
    "\n",
    "## Make individual GSEA plots\n",
    "          local_path2 <- paste0(local_path, 'Individual_GSEA_plot_', ontology, '/')\n",
    "          unlink(local_path2, recursive = T)\n",
    "          dir.create(local_path2)\n",
    "\n",
    "          for (pathway in head(gsea_results$ID, 20)) {\n",
    "               pathway_name <- gsea_results |> as_tibble() |> filter(ID == pathway) |> pull(Description)\n",
    "               anno <- gsea_results[pathway, c(\"NES\", \"pvalue\", \"p.adjust\")]\n",
    "               lab <- paste0(names(anno), \"=\",  round(anno, 4), collapse=\"\\n\")\n",
    "\n",
    "               try(p1 <- enrichplot::gseaplot2(gsea_results, geneSetID = pathway, pvalue_table = FALSE, subplots = 1, base_size = 13,title = paste0(pathway, ' ', pathway_name)))\n",
    "\n",
    "               try(x_position <- ggplot_build(p1)$layout$panel_params[[1]]$x.range[2]*0.75)\n",
    "               try(y_position <- ggplot_build(p1)$layout$panel_params[[1]]$y.range[2]-(ggplot_build(p1)$layout$panel_params[[1]]$y.range[2]-ggplot_build(p1)$layout$panel_params[[1]]$y.range[1])*0.17)\n",
    "\n",
    "               try(p1 <- p1 + annotate(\"text\", x_position, y_position, label = lab, hjust=0, vjust=0, size = 5))\n",
    "               try(p2 <- enrichplot::gseaplot2(gsea_results, geneSetID = pathway, pvalue_table = FALSE, subplots = 2, base_size = 13))\n",
    "               try(p3 <- enrichplot::gseaplot2(gsea_results, geneSetID = pathway, pvalue_table = FALSE, subplots = 3, base_size = 13))\n",
    "\n",
    "               try(cowplot::plot_grid(p1, p2, p3, ncol = 1, rel_heights = c(1.5, 0.5, 1), align = 'v'))    \n",
    "               try(ggsave(paste0('GSEA ',pathway_name , '.pdf'), path = local_path2, height = 10, width = 8))\n",
    "               }\n",
    "          }\n",
    "}\n",
    "\n",
    "GO_functional_analysis <- function (results,  cluster, path='./') {\n",
    "\n",
    "     results <- results[which(duplicated(results$genes) == F),]\n",
    "#    results$entrezid <-  results |> pull(genes) |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID)\n",
    "    #results <- results[which(duplicated(results$entrezid) == F),]\n",
    "\n",
    "\n",
    "####################################### UP ########################################\n",
    "\n",
    "     local_path <- paste0(path, 'GO_functional_analysis_UP_', cluster, '/')\n",
    "     unlink(local_path, recursive = T)\n",
    "     dir.create(local_path)\n",
    "\n",
    "     significant_genes <- results |> filter((padj < 0.05) & (log2FoldChange > 0)) |> arrange(padj) |> pull(genes)\n",
    "     all_genes <- results |> arrange(padj) |> pull(genes)\n",
    "\n",
    "     ######################## ORA ########################\n",
    "\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'ALL')\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'BP')\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'MF')\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'CC')\n",
    "\n",
    "     #################### GSEA ####################\n",
    "\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'ALL')\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'BP')\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'MF')\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'CC')\n",
    "\n",
    "######################################## DOWN ########################################\n",
    "\n",
    "     local_path <- paste0(path, 'GO_functional_analysis_DOWN_', cluster, '/')\n",
    "     unlink(local_path, recursive = T)\n",
    "     dir.create(local_path)\n",
    "\n",
    "     significant_genes <- results |> filter((padj < 0.05) & (log2FoldChange < 0)) |> arrange(padj) |> pull(genes)\n",
    "     all_genes <- results |> arrange(padj) |> pull(genes)\n",
    "\n",
    "     ######################## ORA ########################\n",
    "\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'ALL')\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'BP')\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'MF')\n",
    "     GO_overrepresentation_analysis(significant_genes, all_genes, local_path, ontology = 'CC')\n",
    "\n",
    "     #################### GSEA ####################\n",
    "\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'ALL')\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'BP')\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'MF')\n",
    "     GO_GSEA_analysis(results, local_path, ontology = 'CC')\n",
    "\n",
    "     return()\n",
    "}\n",
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "GO_functional_analysis_cluster_identification <- function (scRNAseq, results, identities = 'seurat_clusters', path='./') {\n",
    "\n",
    "    color_scale <- viridis(n = 4, direction = -1)\n",
    "    options(enrichplot.colours = color_scale)\n",
    "     local_path <- paste0(path, 'Cluster_identification_functional_analysis/')\n",
    "     unlink(local_path, recursive = T)\n",
    "     dir.create(local_path)\n",
    "\n",
    "     all_genes <- Features(scRNAseq[['RNA']]) |>unique()\n",
    "     all_genes_entrezid <- Features(scRNAseq[['RNA']]) |>unique() |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID) |> unique()\n",
    "\n",
    "     mouse_database <- msigdbr(species = 'Mus musculus',category = 'C8') |> dplyr::select(gs_name, entrez_gene)\n",
    "\n",
    "for (cluster in unique(scRNAseq@meta.data |> pull(!!identities))) {\n",
    "\n",
    "####################################### GO ########################################\n",
    "\n",
    "     name <- paste0('Cluster ', cluster, ' - ')\n",
    "\n",
    "     significant_results_cluster <- results |> \n",
    "          filter(cluster == cluster) |>\n",
    "          pull(gene) |>\n",
    "          unique()\n",
    "\n",
    "     GO_overrepresentation_analysis(significant_results_cluster, all_genes, local_path, ontology = 'ALL', minGSSize = 5, maxGSSize = 2000, filename = name)  \n",
    " \n",
    "#################### msigdbr ####################\n",
    "\n",
    "     significant_results_cluster  <-  significant_results_cluster |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID) |> unique() \n",
    "     enrichment_results <- enricher(gene = significant_results_cluster, \n",
    "                    universe = all_genes_entrezid,\n",
    "                    #keyType = \"ENTREZID\",\n",
    "                    #OrgDb = org.Mm.eg.db, \n",
    "                    pAdjustMethod = \"BH\", \n",
    "                    minGSSize    = 5,\n",
    "                    maxGSSize    = 2000,\n",
    "                    qvalueCutoff = 0.05,\n",
    "                    TERM2GENE = mouse_database)\n",
    "                        \n",
    "        enrichment_results_table <- as_tibble(enrichment_results)\n",
    "        write.csv(enrichment_results_table, paste0(local_path, name,'MSigDbr_OverRepresentation_analysis_results_', 'C8','.csv'))\n",
    "\n",
    "        if (nrow(enrichment_results_table) > 0) {\n",
    "            ## Add similarity matrix to the termsim slot of enrichment result\n",
    "            enrichment_results <- enrichplot::pairwise_termsim(enrichment_results)\n",
    "            \n",
    "            enrichment_results_table <- as_tibble(enrichment_results) \n",
    "            write.csv(enrichment_results_table, paste0(local_path, 'MSigDbr_OverRepresentation_analysis_results_', 'C8','.csv'))\n",
    "\n",
    "            dotplot(enrichment_results,\n",
    "                showCategory=50,\n",
    "                title = paste0('MSigDbr Overrepresentation analysis ', 'C8'),\n",
    "                label_format = 60)\n",
    "            ggsave(paste0(name,'MSigDbr_overrepresentation_analysis_dotplot_', 'C8','.pdf'), width = 10, height = 18, path = local_path)\n",
    "\n",
    "            ## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms\n",
    "            emapplot(enrichment_results, showCategory = 60,\n",
    "                cex.params = list(category_node = 0.8, line = 0.3, category_label = 1)) + ggtitle(paste0('Overrepresentation analysis ', 'C8'))\n",
    "            ggsave(paste0(name,'MSigDbr_overrepresentation_analysis_network_', 'C8','.pdf'), width = 14, height = 18, path = local_path)\n",
    "        }\n",
    "     \n",
    "}\n",
    "     return()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "msigdbr_functional_analysis <- function (results,cluster,  path='./') { \n",
    "    \n",
    "    results <- results[which(duplicated(results$genes) == F),]\n",
    "    results$entrezid <-  results |> pull(genes) |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID)\n",
    "    results <- results[which(duplicated(results$entrezid) == F),]\n",
    "\n",
    "        \n",
    "    local_path <- paste0(path, 'MSigDb_functional_analysis_', cluster,'/')\n",
    "    unlink(local_path, recursive = T)\n",
    "    dir.create(local_path)\n",
    "\n",
    "    color_scale <- viridis(n = 4, direction = -1)\n",
    "    options(enrichplot.colours = color_scale)\n",
    "\n",
    "    for (database in c('H', 'C2', 'C3', 'C8')) {\n",
    "\n",
    "        mouse_database <- msigdbr(species = 'Mus musculus',category = database) |> dplyr::select(gs_name, entrez_gene)\n",
    "\n",
    "        #### ORA ####\n",
    "\n",
    "        significant_genes <- filter(results, padj < 0.05) |> arrange(padj) |> pull(entrezid)\n",
    "        all_genes <- results |> arrange(padj) |> pull(entrezid)\n",
    "\n",
    "        enrichment_results <- enricher(gene = significant_genes, \n",
    "                        universe = all_genes,\n",
    "                        #keyType = \"ENTREZID\",\n",
    "                        #OrgDb = org.Mm.eg.db, \n",
    "                        pAdjustMethod = \"BH\", \n",
    "                        minGSSize    = 5,\n",
    "                        maxGSSize    = 500,\n",
    "                        qvalueCutoff = 0.05,\n",
    "                        TERM2GENE = mouse_database)\n",
    "                        \n",
    "        enrichment_results_table <- as_tibble(enrichment_results)\n",
    "        write.csv(enrichment_results_table, paste0(local_path, 'OverRepresentation_analysis_results_', database,'.csv'))\n",
    "\n",
    "        if (nrow(enrichment_results_table) > 0) {\n",
    "            ## Add similarity matrix to the termsim slot of enrichment result\n",
    "            enrichment_results <- enrichplot::pairwise_termsim(enrichment_results)\n",
    "            \n",
    "            enrichment_results_table <- as_tibble(enrichment_results) \n",
    "            write.csv(enrichment_results_table, paste0(local_path, 'OverRepresentation_analysis_results_', database,'.csv'))\n",
    "\n",
    "            dotplot(enrichment_results,\n",
    "                showCategory=50,\n",
    "                title = paste0('Overrepresentation analysis ', database),\n",
    "                label_format = 60)\n",
    "            ggsave(paste0('ORA_dotplot_', database,'.pdf'), width = 10, height = 18, path = local_path)\n",
    "\n",
    "            ## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms\n",
    "            emapplot(enrichment_results, showCategory = 60,\n",
    "                cex.params = list(category_node = 0.8, line = 0.3, category_label = 1)) + ggtitle(paste0('Overrepresentation analysis ', database))\n",
    "            ggsave(paste0('ORA_network_', database,'.pdf'), width = 14, height = 18, path = local_path)\n",
    "        }\n",
    "\n",
    "        #### GSEA ####\n",
    "\n",
    "        fold_changes <- results |> arrange(desc(log2FoldChange)) |> pull(log2FoldChange)\n",
    "        names(fold_changes) <- results |> arrange(desc(log2FoldChange)) |> pull(entrezid)\n",
    "\n",
    "        gsea_results <- GSEA(geneList     = fold_changes,\n",
    "                    # OrgDb        = org.Mm.eg.db,\n",
    "                    # ont          = \"ALL\",\n",
    "                    # keyType = \"SYMBOL\",\n",
    "                    minGSSize    = 5,\n",
    "                    maxGSSize    = 500,\n",
    "                    pvalueCutoff = 0.05,\n",
    "                    verbose      = FALSE,\n",
    "                    TERM2GENE = mouse_database)\n",
    "        \n",
    "        gsea_results_table <- as_tibble(gsea_results)\n",
    "        write.csv(gsea_results_table, paste0(local_path, 'GSEA_analysis_results_', database,'.csv'))\n",
    "\n",
    "        if (nrow(gsea_results_table) > 1) {\n",
    "            gsea_results_table <- as_tibble(gsea_results) \n",
    "            write.csv(gsea_results_table, paste0(local_path, 'GSEA_analysis_results_', database,'.csv'))\n",
    "\n",
    "            dotplot(gsea_results,\n",
    "                showCategory=50,\n",
    "                title = paste0('GSEA analysis ', database),\n",
    "                label_format = 60)\n",
    "            ggsave(paste0('GSEA_dotplot_', database,'.pdf'), width = 10, height = 18, path = local_path)\n",
    "\n",
    "            ## Add similarity matrix to thenes,  termsim slot of enrichment result\n",
    "            gsea_results <- enrichplot::pairwise_termsim(gsea_results)\n",
    "\n",
    "            ## Enrichmap clusters the 50 most significant (by padj) GO terms to visualize relationships between terms\n",
    "            emapplot(gsea_results, showCategory = 60,\n",
    "                cex.params = list(category_node = 0.8, line = 0.3, category_label = 1)) + ggtitle(paste0('GSEA analysis ', database))\n",
    "            ggsave(paste0('GSEA_network_', database,'.pdf'), width = 14, height = 18, path = local_path)\n",
    "        }\n",
    "        \n",
    "    }\n",
    "\n",
    "    \n",
    "    return()\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "pathways_of_interest_analysis <- function (results,  pathways_of_interest, cluster,path='./', FC_threshold = 0.3, p_value_threshold = 0.1, max_overlaps = 1000, label_size = 5, group1, group2, comparison) {\n",
    "\n",
    "    results <- results[which(duplicated(results$genes) == F),]\n",
    "    results$entrezid <-  results |> pull(genes) |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID)\n",
    "    results <- results[which(duplicated(results$entrezid) == F),]\n",
    "        \n",
    "    local_path <- paste0(path, 'pathways_of_interest_', cluster, '/')\n",
    "    unlink(local_path, recursive = T)\n",
    "    dir.create(local_path)\n",
    "\n",
    "\n",
    "    pathways_of_interest <- pathways_of_interest |>\n",
    "        pivot_longer(cols = everything(),values_to = 'genes', names_to = 'pathway', values_drop_na = TRUE) |>\n",
    "        arrange(desc(pathway)) \n",
    "\n",
    "    pathways_of_interest$genes <- pathways_of_interest |> pull(genes) |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID) \n",
    "\n",
    "    pathways_of_interest <- pathways_of_interest |> filter(!is.na(genes) )\n",
    "\n",
    "    #head(pathways_of_interest)\n",
    "\n",
    "    fold_changes <- results |> arrange(desc(log2FoldChange)) |> pull(log2FoldChange)\n",
    "    names(fold_changes) <- results |> arrange(desc(log2FoldChange)) |> pull(entrezid)\n",
    "\n",
    "    for (term_of_interest in unique(pathways_of_interest$pathway)) {\n",
    "\n",
    "        term <- pathways_of_interest |> filter(pathway == term_of_interest)\n",
    "\n",
    "        gsea_results <- GSEA(geneList     = fold_changes,\n",
    "                    # OrgDb        = org.Mm.eg.db,\n",
    "                    # ont          = \"ALL\",\n",
    "                    # keyType = \"SYMBOL\",\n",
    "                    minGSSize    = 4,\n",
    "                    maxGSSize    = 500,\n",
    "                    pvalueCutoff = 1,\n",
    "                    verbose      = FALSE,\n",
    "                    TERM2GENE = term)\n",
    "                    \n",
    "        anno <- gsea_results[term_of_interest, c(\"NES\", \"pvalue\", \"p.adjust\")]\n",
    "        lab <- paste0(names(anno), \"=\",  round(anno, 4), collapse=\"\\n\")\n",
    "\n",
    "        p1 <- enrichplot::gseaplot2(gsea_results, geneSetID = term_of_interest, pvalue_table = FALSE, subplots = 1, base_size = 13,title = term_of_interest)\n",
    "\n",
    "        x_position <- ggplot_build(p1)$layout$panel_params[[1]]$x.range[2]*0.75\n",
    "        y_position <- ggplot_build(p1)$layout$panel_params[[1]]$y.range[2]-(ggplot_build(p1)$layout$panel_params[[1]]$y.range[2]-ggplot_build(p1)$layout$panel_params[[1]]$y.range[1])*0.17\n",
    "\n",
    "        p1 <- p1 + annotate(\"text\", x_position, y_position, label = lab, hjust=0, vjust=0, size = 5)\n",
    "        p2 <- enrichplot::gseaplot2(gsea_results, geneSetID = term_of_interest, pvalue_table = FALSE, subplots = 2, base_size = 13)\n",
    "        p3 <- enrichplot::gseaplot2(gsea_results, geneSetID = term_of_interest, pvalue_table = FALSE, subplots = 3, base_size = 13)     \n",
    "                \n",
    "        cowplot::plot_grid(p1, p2, p3, ncol = 1, rel_heights = c(1.5, 0.5, 1), align = 'v')    \n",
    "        ggsave(paste0('GSEA_', term_of_interest, '.pdf'), path = local_path, height = 10, width = 8)\n",
    "\n",
    "        ########## scatterplot ##########\n",
    "\n",
    "        ## prepare for visualization\n",
    "        results <- results %>% \n",
    "                        \n",
    "                        mutate(                        \n",
    "                            genes_to_label2 = ifelse(entrezid %in% term$genes,  genes ,''),\n",
    "                            add_label = ifelse(entrezid %in% term$genes,  'YES' ,'NO')\n",
    "                            ) |>\n",
    "                        mutate(add_label <- factor(add_label, levels = c('NO', 'YES'))) |>     \n",
    "                        mutate(fold_change_direction = case_when(\n",
    "                                                log2FoldChange>= 0 ~ 'UP',\n",
    "                                                log2FoldChange<= 0 ~ \"DOWN\",\n",
    "                                                TRUE ~ 'NO')) |>\n",
    "                        arrange(add_label)\n",
    "                         \n",
    "                            \n",
    "        my_colors <- c( \"#32228b\", \"gray\")\n",
    "        names(my_colors) <- c(\"YES\", \"NO\")\n",
    "        \n",
    "        # Scatterplot\n",
    "        limx <- results |> pull(paste0('Avg_', group1)) |> max()\n",
    "        limy <- results |> pull(paste0('Avg_', group2)) |> max()\n",
    "        mylims <- max(limx, limy)*5\n",
    "        \n",
    "        results |> \n",
    "            ggplot(aes(x = !!sym(paste0('Avg_', group1)), y = !!sym(paste0('Avg_', group2)), label = genes_to_label2, col = add_label))+\n",
    "                geom_point(size=1.3\n",
    "                ) +\n",
    "                geom_abline(slope = 1, intercept = 0)+\n",
    "                geom_text_repel(\n",
    "                    size=label_size,\n",
    "                    box.padding = 0.2,\n",
    "                    show.legend = FALSE,\n",
    "                    max.overlaps = max_overlaps,\n",
    "                    max.time = 30,\n",
    "                    max.iter = 10000000,\n",
    "                    nudge_x = ifelse(results$fold_change_direction == 'UP', -0.75, 0.75),\n",
    "                    nudge_y = ifelse(results$fold_change_direction == 'UP', 0.75, -0.75),\n",
    "                    aes(segment.size=0.3, segment.alpha=0.4, segment.curvature=0)) +\n",
    "            scale_colour_manual(values=my_colors)+\n",
    "            \n",
    "            \n",
    "            theme(text=element_text(size=20), legend.position=\"none\")+\n",
    "            labs(title=term_of_interest,\n",
    "                        x=paste0('Average Normalized Counts in ', group1),\n",
    "                        y=paste0('Average Normalized Counts in ',  group2))+       \n",
    "            theme_classic(base_size = 28, base_line_size=1) +\n",
    "            theme(legend.position=\"none\", \n",
    "                title = element_text(size=11),\n",
    "                axis.text= element_text(size=10),\n",
    "                axis.title= element_text(size=13))+\n",
    "        scale_x_log10(limits =  c(0.5, mylims))+\n",
    "        scale_y_log10(limits =  c(0.5, mylims))\n",
    "\n",
    "        ggsave(paste0(local_path, 'Pseudobulk scatter ', term_of_interest, '.pdf'))\n",
    "    }\n",
    "    return()\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "source": [
    "### Over representation analysis -- gProfiler2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "gProfiler2_overrepresentation_analysis <- function (significant_genes_FC_ordered, local_path, comparison, filename = '') {\n",
    "\n",
    "    significant_genes_FC_ordered <- list(comparison = significant_genes_FC_ordered)\n",
    "    # Enrichment pathway analysis\n",
    "    enrichment_results <- gost(query = significant_genes_FC_ordered, \n",
    "                    organism = \"mmusculus\", ordered_query = TRUE, \n",
    "                    multi_query = FALSE, significant = TRUE, exclude_iea = FALSE, \n",
    "                    measure_underrepresentation = FALSE, evcodes = FALSE, \n",
    "                    user_threshold = 0.05, correction_method = \"g_SCS\", \n",
    "                    domain_scope = \"annotated\", custom_bg = NULL, \n",
    "                    numeric_ns = \"\", sources = NULL, as_short_link = FALSE, highlight = TRUE)\n",
    "\n",
    "    if (!(is.null(enrichment_results[['result']]) )) {\n",
    "\n",
    "     # Manhattan plot of results\n",
    "     plot  <- gostplot(enrichment_results, capped = T, interactive=T)\n",
    "     htmlwidgets::saveWidget((plot), paste0(local_path, filename, 'Pathway_enrichment_analysis_gprofiler2', '.html'))\n",
    "\n",
    "     # Saving results tablec\n",
    "     enrichment_results <- enrichment_results[['result']] |> dplyr::select(-c('parents'))\n",
    "     write.csv(enrichment_results, paste0(local_path, filename, 'Pathway_enrichment_analysis_gprofiler2', '.csv')) \n",
    "    }\n",
    "    \n",
    "}\n",
    "\n",
    "gProfiler2_functional_analysis <- function (results,  cluster, comparison, path='./', FC_threshold) {\n",
    "\n",
    "     results <- results[which(duplicated(results$genes) == F),]\n",
    "\n",
    "\n",
    "####################################### UP ########################################\n",
    "\n",
    "     local_path <- paste0(path, 'Functional_analysis_UP_gProfiler2_', cluster, '_', comparison, '/')\n",
    "     unlink(local_path, recursive = T)\n",
    "     dir.create(local_path)\n",
    "     print(colnames(results))\n",
    "\n",
    "     significant_genes <- results |> filter((padj < 0.05) & (log2FoldChange > FC_threshold)) |> arrange(padj) |> arrange(desc(log2FoldChange)) |> pull(genes)\n",
    "     \n",
    "     if (length(significant_genes) > 2) {\n",
    "          gProfiler2_overrepresentation_analysis(significant_genes, local_path =  local_path , comparison = comparison, filename = '')\n",
    "\n",
    "     }\n",
    "\n",
    "######################################## DOWN ########################################\n",
    "\n",
    "     local_path <- paste0(path, 'Functional_analysis_DOWN_gProfiler2_', cluster, '_', comparison, '/')\n",
    "     unlink(local_path, recursive = T)\n",
    "     dir.create(local_path)\n",
    "\n",
    "     significant_genes <- results |> filter((padj < 0.05) & (log2FoldChange < -1*FC_threshold)) |> arrange(padj) |> arrange(log2FoldChange) |> pull(genes)\n",
    "\n",
    "     if (length(significant_genes) > 2) {\n",
    "          gProfiler2_overrepresentation_analysis(significant_genes, local_path =  local_path , comparison = comparison, filename = '')\n",
    "     }\n",
    "\n",
    "     return()\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "gProfiler2_functional_analysis_cluster_identification <- function (scRNAseq, results, identities = 'seurat_clusters', path='./') {\n",
    "\n",
    "    color_scale <- viridis(n = 4, direction = -1)\n",
    "    options(enrichplot.colours = color_scale)\n",
    "     local_path <- paste0(path, 'Cluster_identification_functional_analysis_gProfiler2/')\n",
    "     unlink(local_path, recursive = T)\n",
    "     dir.create(local_path)\n",
    "\n",
    "     all_genes <- Features(scRNAseq[['RNA']]) |>unique()\n",
    "     #all_genes_entrezid <- Features(scRNAseq[['RNA']]) |>unique() |> bitr(fromType = 'SYMBOL', toType = 'ENTREZID', OrgDb = 'org.Mm.eg.db', drop = FALSE) |> pull(ENTREZID) |> unique()\n",
    "\n",
    "     #mouse_database <- msigdbr(species = 'Mus musculus',category = 'C8') |> dplyr::select(gs_name, entrez_gene)\n",
    "\n",
    "for (cluster in levels(scRNAseq@meta.data |> pull(!!identities))) {\n",
    "\n",
    "\n",
    "     name <- paste0('Cluster ', cluster, ' - ')\n",
    "\n",
    "     significant_results_cluster <- results |> \n",
    "          filter(cluster == cluster) |>\n",
    "          pull(gene) |>\n",
    "          unique()\n",
    "    \n",
    "     gProfiler2_overrepresentation_analysis(significant_results_cluster, local_path, comparison = 'clusters', filename =  name )\n",
    " \n",
    "\n",
    "     }\n",
    "     return()\n",
    "}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Pseudobulk function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "vscode": {
     "languageId": "r"
    }
   },
   "outputs": [],
   "source": [
    "########## Pseudobulk ##########\n",
    "\n",
    "pseudobulk <- function (scRNAseq, comparison, group1, group2, cluster='all_clusters', path='./', FC_threshold = 0.3, p_value_threshold = 0.05, max_overlaps = 15, label_size = 5, pathways_of_interest = NULL, label_threshold = 100000, distance_from_diagonal_threshold = 0.4, gene_lists_to_plot = NULL, expression_threshold_for_gene_list = 20) {\n",
    "\n",
    "\n",
    "    gene_lists_path <- paste0(path, 'gene_lists/')\n",
    "    figures_path <- paste0(path, 'figures/')\n",
    "\n",
    "    # unlink(gene_lists_path, recursive = T)\n",
    "    # unlink(figures_path, recursive = T)\n",
    "    dir.create(gene_lists_path)\n",
    "    dir.create(figures_path)\n",
    "    print(paste('Cluster',cluster))\n",
    "\n",
    "    Idents(scRNAseq) <- comparison\n",
    "\n",
    "\n",
    "    ########## Continue here ##########\n",
    "\n",
    "    print('number of cells in group 1')\n",
    "    #print(scRNAseq@meta.data |> filter( !!sym(comparison) 'Cluster Annotations'in'Cluster Annotations' c(group1) ) |> nrow())\n",
    "    print(scRNAseq@meta.data |> filter(str_detect( !!as.name(comparison) , group1 )) |> nrow())\n",
    "    print('number of cells in group 2')\n",
    "    print(scRNAseq@meta.data |> filter(str_detect( !!as.name(comparison) , group2 )) |> nrow())\n",
    "    \n",
    "\n",
    "\n",
    "    #Checked there are enough cells\n",
    "    if ((scRNAseq@meta.data |> \n",
    "            filter(str_detect( !!as.name(comparison) , group1 )) |> \n",
    "            nrow()  < 10) | \n",
    "            (scRNAseq@meta.data |> filter(str_detect( !!as.name(comparison) , group2 )) |> nrow()  < 10)) {\n",
    "        DEG_count <- 'Not enough cells'\n",
    "        DEG_UP_count <- 'Not enough cells'\n",
    "        DEG_DOWN_count <- 'Not enough cells'\n",
    "        return(c(all_count=DEG_count, UP_count=DEG_UP_count, DOWN_count=DEG_DOWN_count))\n",
    "    }\n",
    "    \n",
    "    counts <- AggregateExpression(scRNAseq, group.by=c(comparison),\n",
    "                            assays='RNA',\n",
    "                            slot='counts',\n",
    "                            return.seurat=FALSE)\n",
    "\n",
    "    counts <- counts$RNA |> as.data.frame()\n",
    "\n",
    "    # print(counts)\n",
    "    # print(colnames( counts ))\n",
    "    \n",
    "    # Run DE Analysis\n",
    "    #Generate sample level metadata\n",
    "    colData <- data.frame(samples=colnames(counts)) |>\n",
    "                mutate(condition = ifelse(grepl(group1, samples), group1, group2))\n",
    "    \n",
    "    ## Filter\n",
    "    counts <- counts |> mutate(row_sums=rowSums(counts)) |> filter(row_sums >= 10) |> dplyr::select(-row_sums)\n",
    "    \n",
    "    print('Group 1 Length')\n",
    "    print(nrow(colData |> filter(condition == group1)))\n",
    "    print('Group 2 Length')\n",
    "    print(nrow(colData |> filter(condition == group2)))\n",
    "\n",
    "    #Perform DESeq2\n",
    "    if ((length(unique(colData$condition)) != 2 ) | (nrow(colData |> filter(condition == group1)) < 2) | (nrow(colData |> filter(condition == group2)) < 2)) {\n",
    "        DEG_count <- 'Not enough biological replicates per group'\n",
    "        DEG_UP_count <- 'Not enough biological replicates per group'\n",
    "        DEG_DOWN_count <- 'Not enough biological replicates per group'\n",
    "        return(c(all_count=DEG_count, UP_count=DEG_UP_count, DOWN_count=DEG_DOWN_count))\n",
    "    }\n",
    "    \n",
    "    #Create DESeq2 object\n",
    "    \n",
    "    dds <- DESeqDataSetFromMatrix(countData = counts,\n",
    "                            colData = colData,\n",
    "                            design = ~condition)\n",
    "\n",
    "    dds$condition <- factor(dds$condition, levels = c(group1, group2))\n",
    "\n",
    "    \n",
    "\n",
    "    ## DESeq2 QC\n",
    "    rld <- rlog(dds, blind=TRUE) #rlog normalization\n",
    "\n",
    "    DESeq2::plotPCA(rld, ntop=500, intgroup='condition') #PCA\n",
    "    ggsave(filename=paste0('Pseudobulk_PCA_', cluster, '.pdf'), path=figures_path) \n",
    "\n",
    "    #################### Run DESeq2\n",
    "    dds <- DESeq(dds)\n",
    "\n",
    "    #Check the coefficients for the scRNAseq\n",
    "    resultsNames(dds)\n",
    "\n",
    "    #Generate results object\n",
    "    results <- results(dds) |> as.data.frame()\n",
    "    \n",
    "    #Get Normalized Counts\n",
    "    normalized_counts <- counts(dds, normalized = T)\n",
    "    normalized_counts <- normalized_counts |>\n",
    "            as.data.frame() |>\n",
    "            rownames_to_column('genes') |>\n",
    "            as_tibble() |>\n",
    "            rowwise() |>\n",
    "            mutate(\n",
    "                !!paste0('Avg_', group2) := mean(c_across(contains(group2))),\n",
    "                !!paste0('Avg_', group1) := mean(c_across(contains(group1))),\n",
    "            ) |>\n",
    "            ungroup()\n",
    "\n",
    "    #Add gene annotations:\n",
    "    annotations <- read.csv(\"M:/LPD-MIS members' data/Eduard Ansaldo/Bioinformatics/analysis_templates/annotations.csv\")\n",
    "    results <- results |>\n",
    "                    rownames_to_column('genes') |>\n",
    "                    left_join(y= unique(annotations[,c('gene_name', 'description')]),\n",
    "                        by = c('genes' = 'gene_name')) |>\n",
    "                    left_join(y = normalized_counts, by = c('genes' = 'genes'))\n",
    "\n",
    "    head(results)\n",
    "\n",
    "    results_filtered <- filter(results, padj < p_value_threshold & ((!!sym(paste0('Avg_', group2)) > expression_threshold_for_gene_list) | !!sym(paste0('Avg_', group1)) > expression_threshold_for_gene_list) & (log2FoldChange >= FC_threshold | log2FoldChange  <= -1*FC_threshold)) %>% arrange(padj)\n",
    "    results_filtered_UP <- filter(results_filtered, log2FoldChange >  FC_threshold) \n",
    "    results_filtered_DOWN <- filter(results_filtered, log2FoldChange <  FC_threshold)\n",
    "\n",
    "\n",
    "\n",
    "    write.csv(results |> arrange(desc(padj)), file= paste(paste(gene_lists_path, 'ALL_GENES_DEG_Analysis', sep=''), 'pseudobulk', cluster, group2, 'vs', group1, '.csv', sep='_'))\n",
    "    write.csv(results_filtered_UP, file=paste(paste(gene_lists_path, 'DEG_UP', sep=''), 'pseudobulk', cluster, group2, 'vs', group1, '.csv', sep='_'))\n",
    "    write.csv(results_filtered_DOWN, file=paste(paste(gene_lists_path, 'DEG_DOWN', sep=''), 'pseudobulk', cluster, group2, 'vs', group1, '.csv', sep='_'))\n",
    "    \n",
    "\n",
    "    # Return number of DEGs:\n",
    "    DEG_count <- nrow(results_filtered)\n",
    "    DEG_UP_count <- nrow(results_filtered_UP)\n",
    "    DEG_DOWN_count <- nrow(results_filtered_DOWN)\n",
    "\n",
    "\n",
    "    ## prepare for visualization\n",
    "    results_scatter <- results %>% \n",
    "                    drop_na(pvalue) |>\n",
    "                    mutate(\n",
    "                        log10_pval = log10(padj+10^-90)*-1,\n",
    "                        distance_from_diagonal =  (abs((log10(!!sym(paste0('Avg_', group2))+1)) - (log10(!!sym(paste0('Avg_', group1))+1)))/sqrt(2)) ,\n",
    "                        #bottom_limit = (!!sym(paste0('Avg_', group1))-x_intercept^2)/!!sym(paste0('Avg_', group1)),\n",
    "                        #top_limit = bottom_limitp\n",
    "                         )\n",
    "                        \n",
    "    results_scatter <- results_scatter |> mutate(\n",
    "                        genes_to_label_first = ifelse((log2FoldChange >= FC_threshold | log2FoldChange  <= -1*FC_threshold)  & (padj < p_value_threshold) & (distance_from_diagonal > distance_from_diagonal_threshold) & ((!!sym(paste0('Avg_', group2)) > 100) | !!sym(paste0('Avg_', group1)) > 100), genes,NA),\n",
    "                        genes_to_label_second  = ifelse((log2FoldChange >= FC_threshold | log2FoldChange  <= -1*FC_threshold)  & (padj < p_value_threshold) & ((!!sym(paste0('Avg_', group2)) > label_threshold) | !!sym(paste0('Avg_', group1)) > label_threshold) & is.na(genes_to_label_first), genes,NA),\n",
    "                        genes_to_label = ifelse((log2FoldChange >= FC_threshold | log2FoldChange  <= -1*FC_threshold)  & (padj < p_value_threshold) & is.na(genes_to_label_first) & is.na(genes_to_label_second), genes,NA),\n",
    "                        genes_to_label_volcano = ifelse((log2FoldChange >= FC_threshold | log2FoldChange  <= -1*FC_threshold)  & (padj < p_value_threshold), genes,NA),\n",
    "                        diffexpressed = case_when(\n",
    "                            log2FoldChange>=FC_threshold & padj < p_value_threshold  ~ 'UP', \n",
    "                            log2FoldChange<=-1*FC_threshold & padj < p_value_threshold  ~ \"DOWN\",\n",
    "                            TRUE ~ 'NO'))  |>\n",
    "                    mutate(diffexpressed = factor(diffexpressed, levels = c('NO', 'DOWN', 'UP'))) |>\n",
    "                    arrange(diffexpressed)          \n",
    "    \n",
    "    my_colors <- c(\"green4\", \"darkorchid4\", \"gray\")\n",
    "    names(my_colors) <- c(\"DOWN\", \"UP\", \"NO\")\n",
    "\n",
    "    print(head(results_scatter |> filter(diffexpressed != 'NO') |> arrange(desc(distance_from_diagonal)) |> pull(distance_from_diagonal), n = 11))\n",
    "\n",
    "    #ggplot(results, aes(x= !!paste0('Avg_', group1), y = !!paste0('Avg_', group2), label = genes_to_label, col = diffexpressed))+\n",
    "\n",
    "    \n",
    "    # Scatterplot\n",
    "\n",
    "    limx <- results_scatter |> pull(paste0('Avg_', group1)) |> max()\n",
    "    limy <- results_scatter |> pull(paste0('Avg_', group2)) |> max()\n",
    "    mylims <- max(limx, limy)*6\n",
    "       \n",
    "    results_scatter |> \n",
    "        ggplot(aes(x = !!sym(paste0('Avg_', group1)), y = !!sym(paste0('Avg_', group2)), col = diffexpressed))+\n",
    "            geom_point(size=1.3\n",
    "            ) +\n",
    "            geom_abline(slope = 1, intercept = 0)+\n",
    "            geom_text_repel(\n",
    "                size=label_size,\n",
    "                box.padding = 0.35,\n",
    "                show.legend = FALSE,\n",
    "                max.overlaps = 30,\n",
    "                max.time = 10,\n",
    "                max.iter = 10000000,\n",
    "                nudge_x = ifelse(results_scatter$diffexpressed == 'UP', -1, 1),\n",
    "                nudge_y = ifelse(results_scatter$diffexpressed == 'UP', 0.75, -0.75),\n",
    "                aes(label = genes_to_label_first,segment.size=0.3, segment.alpha=0.4, segment.curvature=0)) +\n",
    "            geom_text_repel(\n",
    "                size=label_size,\n",
    "                box.padding = 0.35,\n",
    "                show.legend = FALSE,\n",
    "                max.overlaps = 10,\n",
    "                max.time = 10,\n",
    "                max.iter = 10000000,\n",
    "                nudge_x = ifelse(results_scatter$diffexpressed == 'UP', -1, 1),\n",
    "                nudge_y = ifelse(results_scatter$diffexpressed == 'UP', 0.75, -0.75),\n",
    "                aes(label = genes_to_label_second, segment.size=0.3, segment.alpha=0.4, segment.curvature=0)) +\n",
    "            geom_text_repel(\n",
    "                size=label_size,\n",
    "                box.padding = 0.35,\n",
    "                show.legend = FALSE,\n",
    "                max.overlaps = 10,\n",
    "                max.time = 10,\n",
    "                max.iter = 10000000,\n",
    "                nudge_x = ifelse(results_scatter$diffexpressed == 'UP', -1, 1),\n",
    "                nudge_y = ifelse(results_scatter$diffexpressed == 'UP', 0.75, -0.75),\n",
    "                aes(label = genes_to_label, segment.size=0.3, segment.alpha=0.4, segment.curvature=0)) +\n",
    "        scale_colour_manual(values=my_colors)+\n",
    "        theme(text=element_text(size=20), legend.position=\"none\")+\n",
    "        labs(title=paste('Pseudobulk DEGs in', str_replace(cluster,pattern = '_',replace = ' ') ),\n",
    "                    x=paste0('Average Normalized Counts in ', group1),\n",
    "                    y=paste0('Average Normalized Counts in ',  group2))+\n",
    "        theme_classic(base_size = 28, base_line_size=1) +\n",
    "        theme(legend.position=\"none\", \n",
    "            title = element_text(size=11),\n",
    "            axis.text= element_text(size=10),\n",
    "            axis.title= element_text(size=13))+\n",
    "       scale_x_log10(limits =  c(0.5, mylims))+\n",
    "       scale_y_log10(limits =  c(0.5, mylims))\n",
    "\n",
    "    ggsave(paste0(figures_path, 'Pseudobulk scatter DEG in ', cluster, '.pdf'))\n",
    "\n",
    "\n",
    "    #Filter values which are not significant but with high  FC that would bias the plot visualization\n",
    "    initial_number_of_genes <- nrow(results_scatter)\n",
    "    max_FC_up_significant <- results_scatter %>% filter(diffexpressed != 'NO') %>% dplyr::select(log2FoldChange) %>% max(na.rm = T)\n",
    "    min_FC_up_significant <- results_scatter %>% filter(diffexpressed != 'NO') %>% dplyr::select(log2FoldChange) %>% min(na.rm = T)\n",
    "    if (min_FC_up_significant > -3 | is.na(min_FC_up_significant)) {\n",
    "        min_FC_up_significant <- -3\n",
    "    }  \n",
    "    if (max_FC_up_significant  < 3 | is.na(max_FC_up_significant)) {\n",
    "        max_FC_up_significant <- 3    \n",
    "    }\n",
    "    results_volcano <- results_scatter %>% filter(!(diffexpressed == 'NO' & (log2FoldChange < min_FC_up_significant | log2FoldChange > max_FC_up_significant)))\n",
    "    final_number_of_genes <- nrow(results_volcano)\n",
    "    print(paste('Removed', initial_number_of_genes-final_number_of_genes, 'non-significant genes that would bias the plot visualization'))\n",
    "\n",
    "    # # Volcano Plot\n",
    "    results_volcano |> \n",
    "        arrange(desc(padj)) |>\n",
    "        ggplot(aes(x=log2FoldChange, y=log10_pval, label=genes_to_label_volcano, col=diffexpressed)) +\n",
    "        geom_point(size=1.5) +\n",
    "        geom_text_repel(\n",
    "            size=label_size,\n",
    "            box.padding = 0.35,\n",
    "            show.legend = FALSE,\n",
    "            max.overlaps = max_overlaps,\n",
    "            max.time = 10,\n",
    "            max.iter = 10000000,\n",
    "            aes(segment.size=0.5, segment.alpha=0.8, segment.curvature=0)) +\n",
    "        scale_colour_manual(values=my_colors)+\n",
    "        geom_vline(xintercept=FC_threshold, col=\"lavenderblush2\", linetype=2, size=0.5) +\n",
    "        geom_vline(xintercept=-FC_threshold, col=\"lavenderblush2\", linetype=2, size=0.5) +\n",
    "        geom_hline(yintercept=-1*log10(p_value_threshold), col=\"lavenderblush2\", linetype=2, size=0.5)+\n",
    "        theme(text=element_text(size=20), legend.position=\"none\")+\n",
    "        labs(title=paste('Pseudobulk DEGs in', str_replace(cluster,pattern = '_',replace = ' ') ),\n",
    "                    x=paste('Average log2 FC (', group2, '/', group1, ')', sep=''),\n",
    "                    y= 'Log10 Adj. p-value')+\n",
    "        # coord_cartesian(xlim=c(-1.15, 1.15), ylim = c(0, 95), expand = FALSE)+\n",
    "        theme_classic(base_size = 28, base_line_size=1) +\n",
    "        theme(legend.position=\"none\", \n",
    "            title = element_text(size=15),\n",
    "            axis.text= element_text(size=10),\n",
    "            axis.title= element_text(size=13),\n",
    "            )         +\n",
    "            scale_y_continuous(n.breaks = 8) +\n",
    "            scale_x_continuous(n.breaks = 8)\n",
    "    ggsave(paste0(figures_path, 'Pseudobulk volcano DEG in ', cluster, '.pdf'))\n",
    "\n",
    "########## Overrepresentation analysis ##########\n",
    "\n",
    "gProfiler2_functional_analysis(results,  cluster = cluster, comparison = comparison, path= path , FC_threshold = FC_threshold)\n",
    "\n",
    "\n",
    "if (!is.null(pathways_of_interest)) {\n",
    "    pathways_of_interest_analysis(results = results, pathways_of_interest = pathways_of_interest,  cluster = cluster, path = path, group1 = group1, group2 = group2, comparison = comparison)\n",
    "    }\n",
    "\n",
    "########## Plotting individual genes of interest ##########\n",
    "    if (!is.null(gene_lists_to_plot)) {\n",
    "        for (gene_list in names(gene_lists_to_plot)) {\n",
    "            genes_to_plot <- gene_lists_to_plot[[gene_list]]\n",
    "            \n",
    "            print(genes_to_plot)\n",
    "            \n",
    "        \n",
    "        results_scatter <- results_scatter |> mutate(\n",
    "                            genes_to_label_first = ifelse(genes %in% genes_to_plot, genes, NA),\n",
    "                            diffexpressed = case_when(\n",
    "                                log2FoldChange>=FC_threshold & padj < p_value_threshold & !(genes %in% genes_to_plot) ~ 'UP', \n",
    "                                log2FoldChange<=-1*FC_threshold & padj < p_value_threshold & !(genes %in% genes_to_plot)  ~ \"DOWN\",\n",
    "                                genes %in% genes_to_plot & log2FoldChange >= 0 & padj < p_value_threshold ~ 'INTEREST UP',\n",
    "                                genes %in% genes_to_plot & log2FoldChange < 0 & padj < p_value_threshold ~ 'INTEREST DOWN',\n",
    "                                TRUE ~ 'NO'))  |>\n",
    "                        mutate(diffexpressed = factor(diffexpressed, levels = c('NO', 'DOWN', 'UP', 'INTEREST DOWN', 'INTEREST UP'))) |>\n",
    "                        arrange(diffexpressed)          \n",
    "        \n",
    "        my_colors <- c(\"green4\", \"darkorchid4\", \"gray\", 'green3', 'darkorchid3')\n",
    "        names(my_colors) <- c(\"DOWN\", \"UP\", \"NO\", 'INTEREST DOWN', 'INTEREST UP')\n",
    "\n",
    "        print(head(results_scatter |> filter(diffexpressed != 'NO') |> arrange(desc(distance_from_diagonal)) |> pull(distance_from_diagonal), n = 11))\n",
    "\n",
    "        #ggplot(results, aes(x= !!paste0('Avg_', group1), y = !!paste0('Avg_', group2), label = genes_to_label, col = diffexpressed))+\n",
    "\n",
    "        \n",
    "        # Scatterplot\n",
    "\n",
    "        limx <- results_scatter |> pull(paste0('Avg_', group1)) |> max()\n",
    "        limy <- results_scatter |> pull(paste0('Avg_', group2)) |> max()\n",
    "        mylims <- max(limx, limy)*6\n",
    "        \n",
    "        results_scatter |> \n",
    "            ggplot(aes(x = !!sym(paste0('Avg_', group1)), y = !!sym(paste0('Avg_', group2)), col = diffexpressed))+\n",
    "                geom_point(size=1.3\n",
    "                ) +\n",
    "                geom_abline(slope = 1, intercept = 0)+\n",
    "                geom_text_repel(\n",
    "                    size=label_size,\n",
    "                    box.padding = 0.35,\n",
    "                    show.legend = FALSE,\n",
    "                    max.overlaps = Inf,\n",
    "                    max.time = 10,\n",
    "                    max.iter = 10000000,\n",
    "                    nudge_x = ifelse(results_scatter$diffexpressed == 'UP' | results_scatter$diffexpressed == 'INTEREST UP', -1, 1),\n",
    "                    nudge_y = ifelse(results_scatter$diffexpressed == 'UP' | results_scatter$diffexpressed == 'INTEREST UP', 0.75, -0.75),\n",
    "                    aes(label = genes_to_label_first,segment.size=0.5, segment.alpha=1, segment.curvature=0)) +\n",
    "            \n",
    "            scale_colour_manual(values=my_colors)+\n",
    "            theme(text=element_text(size=20), legend.position=\"none\")+\n",
    "            labs(title=paste('Pseudobulk DEGs in', str_replace(cluster,pattern = '_',replace = ' ') ),\n",
    "                        x=paste0('Average Normalized Counts in ', group1),\n",
    "                        y=paste0('Average Normalized Counts in ',  group2))+\n",
    "            theme_classic(base_size = 28, base_line_size=1) +\n",
    "            theme(legend.position=\"none\", \n",
    "                title = element_text(size=11),\n",
    "                axis.text= element_text(size=10),\n",
    "                axis.title= element_text(size=13))+\n",
    "        scale_x_log10(limits =  c(0.5, mylims))+\n",
    "        scale_y_log10(limits =  c(0.5, mylims))\n",
    "\n",
    "        ggsave(paste0(figures_path, 'Pseudobulk scatter ',gene_list,' in ', group2, ' vs ', group1, '.pdf'))\n",
    "\n",
    "\n",
    "        #Filter values which are not significant but with high  FC that would bias the plot visualization\n",
    "        initial_number_of_genes <- nrow\n",
    "        (results_scatter)\n",
    "        max_FC_up_significant <- results_scatter %>% filter(diffexpressed != 'NO') %>% dplyr::select(log2FoldChange) %>% max(na.rm = T)\n",
    "        min_FC_up_significant <- results_scatter %>% filter(diffexpressed != 'NO') %>% dplyr::select(log2FoldChange) %>% min(na.rm = T)\n",
    "        if (min_FC_up_significant > -3 | is.na(min_FC_up_significant)) {\n",
    "            min_FC_up_significant <- -3\n",
    "        }  \n",
    "        if (max_FC_up_significant  < 3 | is.na(max_FC_up_significant)) {\n",
    "            max_FC_up_significant <- 3    \n",
    "        }\n",
    "        results_volcano <- results_scatter %>% filter(!(diffexpressed == 'NO' & (log2FoldChange < min_FC_up_significant | log2FoldChange > max_FC_up_significant)))\n",
    "        final_number_of_genes <- nrow(results_volcano)\n",
    "        print(paste('Removed', initial_number_of_genes-final_number_of_genes, 'non-significant genes that would bias the plot visualization'))\n",
    "\n",
    "        # # Volcano Plot\n",
    "        results_volcano |> \n",
    "            #arrange(desc(padj)) |>\n",
    "            ggplot(aes(x=log2FoldChange, y=log10_pval, label=genes_to_label_first, col=diffexpressed)) +\n",
    "            geom_point(size=1.5) +\n",
    "            geom_text_repel(\n",
    "                size=label_size,\n",
    "                box.padding = 0.35,\n",
    "                show.legend = FALSE,\n",
    "                max.overlaps = max_overlaps,\n",
    "                max.time = 10,\n",
    "                max.iter = 10000000,\n",
    "                nudge_x = ifelse(results_scatter$diffexpressed == 'UP' | results_scatter$diffexpressed == 'INTEREST UP', 3 ,-3),\n",
    "                nudge_y = ifelse(results_scatter$diffexpressed == 'UP', 3, 3),\n",
    "                #nudge_x = (5),\n",
    "                aes(segment.size=0.3, segment.alpha=1, segment.curvature=0)) +\n",
    "            scale_colour_manual(values=my_colors)+\n",
    "            geom_vline(xintercept= FC_threshold, col=\"lavenderblush2\", linetype=2, size=0.5) +\n",
    "            geom_vline(xintercept=-FC_threshold, col=\"lavenderblush2\", linetype=2, size=0.5) +\n",
    "            geom_hline(yintercept=-1*log10(p_value_threshold), col=\"lavenderblush2\", linetype=2, size=0.5)+\n",
    "            theme(text=element_text(size=20), legend.position=\"none\")+\n",
    "            labs(title=paste('Pseudobulk DEGs in', str_replace(cluster,pattern = '_',replace = ' ') ),\n",
    "                        x=paste('Average log2 FC (', group2, '/', group1, ')', sep=''),\n",
    "                        y= 'Log10 Adj. p-value')+\n",
    "            # coord_cartesian(xlim=c(-1.15, 1.15), ylim = c(0, 95), expand = FALSE)+\n",
    "            theme_classic(base_size = 28, base_line_size=1) +\n",
    "            theme(legend.position=\"none\", \n",
    "                title = element_text(size=15),\n",
    "                axis.text= element_text(size=10),\n",
    "                axis.title= element_text(size=13),\n",
    "                )         +\n",
    "                scale_y_continuous(n.breaks = 8) +\n",
    "                scale_x_continuous(n.breaks = 8)\n",
    "        ggsave(paste0(figures_path, 'Pseudobulk volcano ',gene_list,' in ', group2, ' vs ', group1, '.pdf'))\n",
    " \n",
    "    }}\n",
    "\n",
    "     return(c(all_count=DEG_count, UP_count=DEG_UP_count, DOWN_count=DEG_DOWN_count))\n",
    "   \n",
    "}                            "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.4.1"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
